src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -245,12 +246,116 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
+            {
+                return;
+            }
+
+            var (leftNames, leftInferred, leftLocations) = getDetails(left);
+            bool leftNoInferredNames = leftInferred.IsDefault;
+            bool leftNoNames = leftNames.IsDefault;
+
+            var (rightNames, rightInferred, rightLocations) = getDetails(right);
+            bool rightNoInferredNames = rightInferred.IsDefault;
+            bool rightNoNames = rightNames.IsDefault;
+
+            if (leftNoNames && rightNoNames)
+            {
+                return;
+            }
+
+            Debug.Assert(leftNoNames || rightNoNames || leftNames.Length == rightNames.Length);
+
+            // The only case where the right side is not preferred is when we have a tuple literal on the left, but not the right
+            bool isRightBestSide = !(leftIsTupleLiteral && !rightIsTupleLiteral);
+
+            int length = leftNoNames ? rightNames.Length : leftNames.Length;
+            for (int i = 0; i < length; i++)
+            {
+                string leftName = leftNoNames ? null : leftNames[i];
+                bool leftWasInferred = leftNoInferredNames ? false : leftInferred[i];
+
+                string rightName = rightNoNames ? null : rightNames[i];
+                bool rightWasInferred = rightNoInferredNames ? false : rightInferred[i];
+
+                bool different = string.CompareOrdinal(rightName, leftName) != 0;
+
+                bool leftComplaint = leftIsTupleLiteral && leftName != null && !leftWasInferred && (rightNoNames || different);
+                bool rightComplaint = rightIsTupleLiteral && rightName != null && !rightWasInferred && (leftNoNames || different);
+
+                if (!leftComplaint && !rightComplaint)
+                {
+                    // No complaints, let's move on
+                    continue;
+                }
+
+                BoundExpression location = null;
+                string complaintName = null;
+                if (leftComplaint && rightComplaint)
+                {
+                    // When in doubt, we'll complain on the right side
+                    location = isRightBestSide ? rightLocations[i] : leftLocations[i];
+                    complaintName = isRightBestSide ? rightName : leftName;
+                }
+                else if (rightComplaint)
+                {
+                    location = rightLocations[i];
+                    complaintName = rightName;
+                }
+                else if (leftComplaint)
+                {
+                    location = leftLocations[i];
+                    complaintName = leftName;
+                }
+
+                diagnostics.Add(ErrorCode.WRN_TupleBinopLiteralNameMismatch, location.Syntax.Parent.Location, complaintName);
+            }
+
+            (ImmutableArray<string> namesOpt, ImmutableArray<bool> inferredOpt, ImmutableArray<BoundExpression> locationsOpt) getDetails(BoundExpression expr)
```

Comment:
Consider using uppercase for first character in tuple element names. Same comment for other methods.